### PR TITLE
Add a "HTML" link style; Click icon to copy

### PR DESCRIPTION
**Pull request by herrkaefer**
With the added "HTML" link style the selected text is copied with its original HTML format, plus a link appended at the end. The purpose is to preserve the text format rather than just copying the text string. 

Explanation of motivation: When copying text fragment for note taking, I want both the text and the link be placed in the note. The "rich" style actually provides this, but seems not perfect: 1) The original format is lost, but many apps can recognize the HTML during copy and display it pretty; 2) The link is applied to the whole text, which sometimes seems annoying especially when you want to select part of the copied text.

I noticed that the URL in the rich style has some encoding issue: the spaces are still space which should be %20. Didn't figure out a way to fix it. So in the HTML style, the link is added as a markdown style: `[🔗 ](url)`.

### CODE REVIEW

File: background_script.js [hunk 1]
```diff
@@ -166,5 +166,18 @@
...
+    if (!SUPPORTS_TEXT_FRAGMENTS) {
+      await polyfillTextFragments();
+    }
+    await injectContentScripts([
+      'prepare.js',
+      'fragment-generation-utils.js',
+      'content_script.js',
+    ]);
+    startProcessing(tab);
+  });
```
Comment (tomayac) [inline background_script.js:181]: Why is this needed?

File: content_script.js [hunk 2]
```diff
@@ -104,14 +104,29 @@
...
+          var html = "";
+          if (selection.rangeCount) {
+              var container = document.createElement("div");
+              for (var i = 0, len = selection.rangeCount; i < len; ++i) {
+                  container.appendChild(selection.getRangeAt(i).cloneContents());
+              }
+              html = container.innerHTML;
+          }
+          clipboardItems['text/html'] = new Blob(
+              [`${html} [🔗](${url})`],
```
Comment (tomayac) [inline content_script.js:125]: This looks like Markdown to me, not HTML.

**Review by tomayac — COMMENTED** [review]
I'm not sure this is universally useful. While it may suit your personal use case (for example, by including the `🔗` emoji), other users may have different preferences.

File: background_script.js (hunk 1 above)
Comment (herrkaefer) [reply background_script.js:181]: This is added to enable "clicking the icon to copy". Basically a copy of the same operations for corresponding click from the context menu.

File: content_script.js (hunk 2 above)
Comment (herrkaefer) [reply content_script.js:125]: Explained here in the first comment:

> I noticed that the URL in the rich style has some encoding issue: the spaces are still space which should be %20. Didn't figure out a way to fix it. So in the HTML style, the link is added as a markdown style.

(I am not sure if someone has reported the URL encoding issue before...)

**Comment by herrkaefer** [comment]
> I'm not sure this is universally useful. While it may suit your personal use case (for example, by including the `🔗` emoji), other users may have different preferences.

@tomayac Thanks for your comments! The modifications are all based on my personal need, and they makes this tool much better than the default behavior, i.e. only copy the link of the text. The 🔗 was a replace of the original style: "Some text [link]" I was wondering a while between them.

Please feel free to ignore this pull request if you feel uncomfortable with the customization. Thanks for your work.

File: background_script.js (hunk 1 above)
Comment (tomayac) [reply background_script.js:181]: I see. In this case, can you please factor out the code in https://github.com/GoogleChromeLabs/link-to-text-fragment/blob/27728509b02305e91d0483fa8774c03796bafe6f/background_script.js#L121-L131 into a function and call it from both locations, `browser.contextMenus.onClicked` and `browser.browserAction.onClicked` (note that I have re-name-spaced `chrome` to `browser`.

File: content_script.js (hunk 2 above)
Comment (tomayac) [reply content_script.js:125]: Thanks. Could you then please fix the encoding issue so a proper HTML link can be used. See https://github.com/GoogleChromeLabs/text-fragments-polyfill/issues/115.

**Comment by tomayac** [comment]
Left some more comments. I think we can add this as a third alternative link option:

- Raw: copies `https://example.com/#:~:text=Example%20Domain,-This%20domain%20is`.
- Rich: copies [`Example Domain`](https://example.com/#:~:text=Example%20Domain,-This%20domain%20is).
- Rich plus raw: copies [`Example Domain`](https://example.com/#:~:text=Example%20Domain,-This%20domain%20is)`🔗 https://example.com/#:~:text=Example%20Domain,-This%20domain%20is`.

File: background_script.js (hunk 1 above)
Comment (herrkaefer) [reply background_script.js:181]: Nice refactoring suggestion! Will do this.

File: content_script.js (hunk 2 above)
Comment (herrkaefer) [reply content_script.js]: Looks like this has been fixed in e0b0a85c8a169da67a09752abd2d992cbdeac7fc

File: background_script.js (hunk 1 above)
Comment (herrkaefer) [reply background_script.js]: Done in 8010cf8bb91af6ca05c3a98762981120f4682a20

**Comment by herrkaefer** [comment]
"HTML" has been renamed to "rich plus raw" in 582a8cbb9e1de8713b70ca824013f031ff8a3081. 

Applying `encodeURI` on the `url` seems to have fixed the encoding issue, by which I mean it has at least passed a few testings on my side. Now "rich" and "rich plus raw" both copy the link as HTML (as `a` element).

File: background_script.js [hunk 3]
```diff
@@ -166,5 +156,20 @@
       );
     });
   };
+
+  const onCopy = async (info, tab) => {
+    console.log("on Copy: clicked");
```
Comment (tomayac) [inline background_script.js:161]:
```suggestion
```

File: content_script.js [hunk 4]
```diff
@@ -104,14 +104,29 @@
...
+                  selection.toString()}}}`,
             ],
             {
               type: 'text/rtf',
             }
           );
           */
         }
+        else if (linkStyle === 'rich_plus_raw') {
+          var html = "";
```
Comment (tomayac) [inline content_script.js:116]:
```suggestion
          let html = '';
```

File: content_script.js [hunk 5]
```diff
@@ -104,14 +104,29 @@
...
+          if (selection.rangeCount) {
+              var container = document.createElement("div");
```
Comment (tomayac) [inline content_script.js:118]:
```suggestion
              const container = document.createElement('div');
```

File: content_script.js [hunk 6]
```diff
@@ -104,14 +104,29 @@
...
+              for (var i = 0, len = selection.rangeCount; i < len; ++i) {
```
Comment (tomayac) [inline content_script.js:119]:
```suggestion
              for (let i = 0, len = selection.rangeCount; i < len; ++i) {
```

File: content_script.js [hunk 7]
```diff
@@ -104,14 +104,29 @@
...
+                  container.appendChild(selection.getRangeAt(i).cloneContents());
+              }
+              html = container.innerHTML;
+          }
+          clipboardItems['text/html'] = new Blob(
+              [`${html} <a href="${encodeURI(url)}">🔗</a>`],
```
Comment (tomayac) [inline content_script.js:125]:
```suggestion
              [`${html} 🔗 <a href="${encodeURI(url)}">${encodeURI(url)}</a>`],
```

I think this is what you meant?!

**Review by tomayac — CHANGES_REQUESTED** [review]
Getting there…

File: content_script.js (hunk 7 above)
Comment (herrkaefer) [reply content_script.js:125]: Unfortunately this change is not what I wanted. To take my use case for instance, I copy a few sentences or a whole paragraph sometimes and the URL could get very long in this case, and it looks concise if shown as a link of the emoji. 

Example:

I want to see:

Democrats have so far used the hearings to praise Brown -- who would be the first Black woman to serve as a Supreme Court justice -- as an exceptionally qualified, trail-blazing nominee whose depth and breadth of experience, including as a federal public defender, would add a valuable and unique perspective to the bench. [🔗](https://www.cnn.com/2022/03/22/politics/ketanji-brown-jackson-confirmation-hearings-day-two/index.html#:~:text=Democrats%20have%20so,to%20the%20bench.)

instead of

Democrats have so far used the hearings to praise Brown -- who would be the first Black woman to serve as a Supreme Court justice -- as an exceptionally qualified, trail-blazing nominee whose depth and breadth of experience, including as a federal public defender, would add a valuable and unique perspective to the bench.  🔗 [https://www.cnn.com/2022/03/22/politics/ketanji-brown-jackson-confirmation-hearings-day-two/index.html#:~:text=Democrats%20have%20so,to%20the%20bench.]( [🔗](https://www.cnn.com/2022/03/22/politics/ketanji-brown-jackson-confirmation-hearings-day-two/index.html#:~:text=Democrats%20have%20so,to%20the%20bench.))

Comment (herrkaefer) [reply content_script.js:125]: Another example to show that the text itself is with HTML style:

Want to get and see:

President Joe Biden's Supreme Court nominee, [Ketanji Brown Jackson, ](https://www.cnn.com/2022/02/02/politics/ketanji-brown-jackson-profile/index.html)defended her judicial record Tuesday as she faced intense questioning from Republican senators during the second day of her historic confirmation hearings. [🔗](https://www.cnn.com/2022/03/22/politics/ketanji-brown-jackson-confirmation-hearings-day-two/index.html#:~:text=president%20joe%20biden's%20supreme%20court%20nominee%2C%20ketanji%20brown%20jackson%2C%20defended%20her%20judicial%20record%20tuesday%20as%20she%20faced%20intense%20questioning%20from%20republican%20senators%20during%20the%20second%20day%20of%20her%20historic%20confirmation%20hearings.)

instead of

President Joe Biden's Supreme Court nominee, [Ketanji Brown Jackson, ](https://www.cnn.com/2022/02/02/politics/ketanji-brown-jackson-profile/index.html)defended her judicial record Tuesday as she faced intense questioning from Republican senators during the second day of her historic confirmation hearings. 🔗 [https://www.cnn.com/2022/03/22/politics/ketanji-brown-jackson-confirmation-hearings-day-two/index.html#:~:text=president%20joe%20biden's%20supreme%20court%20nominee%2C%20ketanji%20brown%20jackson%2C%20defended%20her%20judicial%20record%20tuesday%20as%20she%20faced%20intense%20questioning%20from%20republican%20senators%20during%20the%20second%20day%20of%20her%20historic%20confirmation%20hearings.](https://www.cnn.com/2022/03/22/politics/ketanji-brown-jackson-confirmation-hearings-day-two/index.html#:~:text=president%20joe%20biden's%20supreme%20court%20nominee%2C%20ketanji%20brown%20jackson%2C%20defended%20her%20judicial%20record%20tuesday%20as%20she%20faced%20intense%20questioning%20from%20republican%20senators%20during%20the%20second%20day%20of%20her%20historic%20confirmation%20hearings.)

Comment (herrkaefer) [reply content_script.js:125]: Ha this is what I was thinking. But the options could be very personal choices like emoji, Link, LINK, or link?

How about 1) 🔗 2) [source] 3) [link]? I prefer to use the square brackets to discriminate it from the text and use all lowercases.

Comment (tomayac) [reply content_script.js:125]: Think about international users as well. Thinking about it, I would ideally allow the user to choose:

- Quotes and quote style 
- Parentheses and parentheses style  around link word
- Link word

Two examples:

- “Bla bla” [Source]
- «Bla bla» (#)

Comment (herrkaefer) [reply content_script.js:125]: Good points! Quotes around the text looks nice for some use cases, but it doesn't make sense to me when the HTML text include multiple lines, which I usually copy. I would try to add the option to let user input any preferred "link text", and put the URL behind it.

Comment (herrkaefer) [reply content_script.js]: Link text options are added in the setting page to let the users either choose a preset or input their own favorite word. e6e83376abcc6cda6f5070c101f832605994213a @tomayac Please have a try!

File: options.js [hunk 8]
```diff
@@ -1,26 +1,97 @@
...
+
+    link_text_option_1.checked = items.linkText === document.querySelector('.link_text_option_1').textContent;
+    link_text_option_2.checked = items.linkText === document.querySelector('.link_text_option_2').textContent;
+    link_text_option_3.checked = items.linkText === document.querySelector('.link_text_option_3').textContent;
+    if (!link_text_option_1.checked && !link_text_option_2.checked && !link_text_option_3.checked) {
+        link_text_option_4.checked = true;
+        link_text_input.value = items.linkText;
+    }
+
+    enableLinkTextOptions(rich_plus_raw.checked);
```
Comment (tomayac) [inline options.js:42]: Here and throughout: please use `lowerCamelCase` variable names as per the [Google JS style guide](https://google.github.io/styleguide/jsguide.html#naming-local-variable-names). Thanks!

File: options.js [hunk 9]
```diff
@@ -1,26 +1,97 @@
...
-    browser.i18n.getMessage('link_copy_style');
+  document.querySelector('h1').textContent = browser.i18n.getMessage('link_copy_style');
   document.querySelector('.rich').textContent = browser.i18n.getMessage('rich');
   document.querySelector('.raw').textContent = browser.i18n.getMessage('raw');
+  document.querySelector('.rich_plus_raw').textContent = browser.i18n.getMessage('rich_plus_raw');
+  document.querySelector('h4').textContent = browser.i18n.getMessage('link_text');
+  document.querySelector('.link_text_option_1').textContent = browser.i18n.getMessage('link_text_option_1');
+  document.querySelector('.link_text_option_2').textContent = browser.i18n.getMessage('link_text_option_2');
+  document.querySelector('.link_text_option_3').textContent = browser.i18n.getMessage('link_text_option_3');
+  document.querySelector('.link_text_option_4').textContent = browser.i18n.getMessage('link_text_option_4');
```
Comment (tomayac) [inline options.js:10]: Here and throughout: please use `-` as the separator for class names per the [Google CSS style guide](https://google.github.io/styleguide/htmlcssguide.html#ID_and_Class_Name_Delimiters). Thanks!

**Review by tomayac — CHANGES_REQUESTED** [review]
Just style guide changes now.

File: options.js (hunk 8 above)
Comment (herrkaefer) [reply options.js]: Done in af62d525855da94d9bbc8b0bb3a751c290660d4d

File: options.js (hunk 9 above)
Comment (herrkaefer) [reply options.js]: Done in af62d525855da94d9bbc8b0bb3a751c290660d4d

**Comment by tomayac** [comment]
@herrkaefer Just wanted to let you know that this PR is still on my radar. I simply need to deal with an urgent deadline for an event. Thanks for your patience.

**Comment by herrkaefer** [comment]
@tomayac Thanks for your attention! Would like to update if you have more comments.
